File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/account/application/command/DepositByPaymentService.java
package com.example.accountservice.account.application.command;

import com.example.accountservice.account.domain.Account;
import com.example.accountservice.account.domain.AccountFindQuery;
import com.example.accountservice.account.domain.AccountRepository;
import com.example.accountservice.account.domain.TransactionType;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

/**
 * The reaction use case for both of the Payment BC's {@code payment.cancelled.v1}
 * (payment-cancellation compensating credit) and {@code refund.approved.v1} (refund-approval
 * credit) Integration Events — both events are the same action, "reverse an already-debited
 * amount," differing only in referenceId (paymentId or refundId), so a single Command is reused.
 *
 * <p>Idempotency uses Level 2 Ledger for the same reason as {@link WithdrawByPaymentService}. The
 * reason the shared Outbox drain component is not injected via the constructor is also the same as
 * {@link WithdrawByPaymentService} (avoiding a circular bean dependency — this service is only ever
 * called from inside an in-progress drain loop, via an {@code OutboxEventHandler} implementation).
 */
@Service
@RequiredArgsConstructor
public class DepositByPaymentService {

    private final AccountRepository accountRepository;

    public void deposit(DepositByPaymentCommand command) {
        boolean alreadyProcessed =
                accountRepository.hasTransactionWithReference(
                        command.referenceId(), TransactionType.DEPOSIT);
        if (alreadyProcessed) {
            return;
        }

        Account account =
                accountRepository
                        .findAccounts(new AccountFindQuery(0, 1, command.accountId(), null, null))
                        .accounts()
                        .stream()
                        .findFirst()
                        .orElse(null);
        if (account == null) {
            return;
        }

        account.deposit(command.amount(), command.referenceId());
        accountRepository.saveAccount(account);
    }
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/account/application/command/DepositByPaymentService.kt
package com.example.accountservice.account.application.command

import com.example.accountservice.account.domain.AccountFindQuery
import com.example.accountservice.account.domain.AccountRepository
import com.example.accountservice.account.domain.TransactionType
import org.springframework.stereotype.Service

/**
 * The reaction use case for both the Payment BC's `payment.cancelled.v1` (cancellation compensation
 * credit) and `refund.approved.v1` (refund-approval credit) Integration Events — both events perform the
 * same action, "reversing an amount already deducted," differing only in referenceId (paymentId or
 * refundId), so a single command is reused for both.
 *
 * Idempotency uses a Level 2 Ledger for the same reason as [WithdrawByPaymentService]. This Service also
 * does not call `outboxRelay.processPending()`, for the same reason as [WithdrawByPaymentService].
 */
@Service
class DepositByPaymentService(
    private val accountRepository: AccountRepository,
) {
    fun deposit(command: DepositByPaymentCommand) {
        val alreadyProcessed =
            accountRepository.hasTransactionWithReference(command.referenceId, TransactionType.DEPOSIT)
        if (alreadyProcessed) return

        val (accounts, _) =
            accountRepository.findAccounts(AccountFindQuery(page = 0, take = 1, accountId = command.accountId))
        val account = accounts.firstOrNull() ?: return

        account.deposit(command.amount, command.referenceId)
        accountRepository.saveAccount(account)
    }
}
